simulator/folder_test.go

```diff
@@ -645,3 +646,223 @@ func TestPlaceVmsXCluster(t *testing.T) {
+
+		cfgSpec := types.VirtualMachineConfigSpec{}
+
+		tests := []struct {
+			name         string
+			poolMoRefs   []types.ManagedObjectReference
+			configSpec   types.VirtualMachineConfigSpec
+			relocateSpec *types.VirtualMachineRelocateSpec
+			vmMoRef      *types.ManagedObjectReference
+			expectedErr  string
+		}{
+			{
+				"relocate without any resource pools",
+				nil,
+				cfgSpec,
+				&types.VirtualMachineRelocateSpec{},
+				&vmMoRef,
+				"InvalidArgument",
+			},
+			{
+				"relocate without a relocate spec",
+				poolMoRefs,
+				cfgSpec,
+				nil,
+				&vmMoRef,
+				"InvalidArgument",
+			},
+			{
+				"relocate without a vm in the placement spec",
+				poolMoRefs,
+				cfgSpec,
+				&types.VirtualMachineRelocateSpec{},
+				nil,
+				"InvalidArgument",
+			},
+			{
+				"relocate with a non-existing vm in the placement spec",
+				poolMoRefs,
+				cfgSpec,
+				&types.VirtualMachineRelocateSpec{},
+				&types.ManagedObjectReference{
+					Type:  "VirtualMachine",
+					Value: "fake-vm-999",
+				},
+				"InvalidArgument",
+			},
+			{
+				"relocate with an empty relocate spec",
+				poolMoRefs,
+				cfgSpec,
+				&types.VirtualMachineRelocateSpec{},
+				&vmMoRef,
+				"",
+			},
+		}
+
+		for testNo, test := range tests {
+			test := test // assign to local var since loop var is reused
+
+			placeVmsXClusterSpec := types.PlaceVmsXClusterSpec{
+				ResourcePools: test.poolMoRefs,
+				PlacementType: string(types.PlaceVmsXClusterSpecPlacementTypeRelocate),
+			}
+
+			placeVmsXClusterSpec.VmPlacementSpecs = []types.PlaceVmsXClusterSpecVmPlacementSpec{{
+				ConfigSpec:   test.configSpec,
+				Vm:           test.vmMoRef,
+				RelocateSpec: test.relocateSpec,
+			}}
+
+			folder := object.NewRootFolder(c)
+			res, err := folder.PlaceVmsXCluster(ctx, placeVmsXClusterSpec)
+
+			if err == nil && test.expectedErr != "" {
+				t.Fatalf("Test %v: expected error %q, received nil", testNo, test.expectedErr)
+			} else if err != nil &&
+				(test.expectedErr == "" || !strings.Contains(err.Error(), test.expectedErr)) {
+				t.Fatalf("Test %v: expected error %q, received %v", testNo, test.expectedErr, err)
+			}
+
+			if err == nil {
+				if len(res.PlacementInfos) != len(placeVmsXClusterSpec.VmPlacementSpecs) {
+					t.Errorf("%d PlacementInfos vs %d VmPlacementSpecs", len(res.PlacementInfos), len(placeVmsXClusterSpec.VmPlacementSpecs))
+				}
+			}
+		}
+	}, vpx)
+}
+
+func TestPlaceVmsXClusterReconfigure(t *testing.T) {
+	vpx := VPX()
+	vpx.Cluster = 3
+
+	Test(func(ctx context.Context, c *vim25.Client) {
+		finder := find.NewFinder(c, true)
+		datacenter, err := finder.DefaultDatacenter(ctx)
+		if err != nil {
+			t.Fatalf("failed to get default datacenter: %v", err)
+		}
+		finder.SetDatacenter(datacenter)
+
+		cluster := Map.Any("ClusterComputeResource").(*ClusterComputeResource)
+		pool := cluster.ResourcePool
+		host := Map.Get(cluster.Host[0]).(*HostSystem)
+		vm := Map.Get(host.Vm[0]).(*VirtualMachine)
```

Comment:
Looks like this could be an issue, if the random/Any ClusterComputeResource doesn't have any VMs placed on it.
Could do the following instead:
- disable standalone host
- choose random/Any VirtualMachine, then derive host + cluster from that vm
```diff
diff --git a/simulator/folder_test.go b/simulator/folder_test.go
index 42dc36be..706c2966 100644
--- a/simulator/folder_test.go
+++ b/simulator/folder_test.go
@@ -762,6 +762,7 @@ func TestPlaceVmsXClusterRelocate(t *testing.T) {
 func TestPlaceVmsXClusterReconfigure(t *testing.T) {
 	vpx := VPX()
 	vpx.Cluster = 3
+	vpx.Host = 0 // no standalone hosts
 
 	Test(func(ctx context.Context, c *vim25.Client) {
 		finder := find.NewFinder(c, true)
@@ -771,10 +772,10 @@ func TestPlaceVmsXClusterReconfigure(t *testing.T) {
 		}
 		finder.SetDatacenter(datacenter)
 
-		cluster := Map.Any("ClusterComputeResource").(*ClusterComputeResource)
+		vm := Map.Any("VirtualMachine").(*VirtualMachine)
+		host := Map.Get(*vm.Runtime.Host).(*HostSystem)
+		cluster := Map.Get(*host.Parent).(*ClusterComputeResource)
 		pool := cluster.ResourcePool
-		host := Map.Get(cluster.Host[0]).(*HostSystem)
-		vm := Map.Get(host.Vm[0]).(*VirtualMachine)
 
 		var poolMoRefs []types.ManagedObjectReference
 		poolMoRefs = append(poolMoRefs, pool.Reference())
```

Reply:
Yeah, exactly.  Once I disabled standalone host, I was able to traverse up to CCR.  My latest has that change.